misk-aws/src/main/kotlin/misk/jobqueue/sqs/SqsJob.kt

```diff
@@ -55,6 +58,36 @@ internal class SqsJob(
     metrics.jobsDeadLettered.labels(queueName.value, queueName.value).inc()
   }
 
+  /**
+   *  Assign a visibility timeout for some duration X making the job invisible to the consumers for
+   *  that duration. With every subsequent retry the duration becomes longer until it hits the max
+   *  value of 10hrs.
+   */
+   fun setVisibilityTimeout(maxRetryCount: Int = 10) {
+    val visibilityTime = calculateVisibilityTimeOut(maxRetryCount)
+    queue.call { client ->
+      client.changeMessageVisibility(
+        ChangeMessageVisibilityRequest()
+          .withQueueUrl(queue.url)
+          .withReceiptHandle(message.receiptHandle)
+          .withVisibilityTimeout(visibilityTime)
+      )
+    }
+    metrics.visibilityTime.labels(queueName.value, queueName.value).set(visibilityTime.toDouble())
+  }
+
+  private fun calculateVisibilityTimeOut(maxReceiveCount: Int): Int {
+    val currentReceiveCount = attributes["ApproximateReceiveCount"]?.toInt() ?: 1
+
+    val consecutiveRetryCount = (currentReceiveCount + 1).coerceAtMost(maxReceiveCount)
+    val backoff = 2.0.pow((consecutiveRetryCount - 1).toDouble()).toLong()
+    val maxDelay = 10 * 60 * 60L // We are limited with 12hrs so let's just limit it to 10hrs
+
+    val backoffWithJitter = maxDelay.coerceAtMost((backoff / 2 + Random.nextLong(0, backoff / 2)))
```

Comment:
Could we add a test to ensure we're not going over the 10 hours?

Reply:
Hmm this one would be a bit tricky to make - current job setup actually makes a call to the SQS.
I also don't want to expose this function making it public. I tried playing around with mocks but that didn't work well either.

Reply:
You could modify the visibility timeout in the test itself, then pull from AWS again. The receive count will still be incremented, so when it fails again you'll get the next backoff amount

Reply:
hmm i tried doing that but that would involve making receiptHandle on the sqs job visible, fetching the job queue url from the resolved queues and bunch of other changes. in order to make things simpler - i just moved the function into an object and added quick unit test